feat(manager/composer): support git-tags hostRules for github.com when updating artifacts

etremblay opened the pull request:
## Changes
Take 2 of #16193 considering 

When generating the COMPOSER_AUTH json, handle special case for github host rule.

```
"hostRules": [
    {
      "hostType": "git-tags", 
      "matchHost": "github.com",
      "encrypted": {
        "token": "GitHub Personal Access Token"
      }
    },
]
```

## Context

- fixes #10691 
- ref #17778

We cannot use the `github` hostType because it break the update of the dependancy dashboard.

Also, I found that git-tags hostRules where used when doing dependancy lookup so this fix everything related to private github php repositorie

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

etremblay commented on lib/modules/manager/composer/artifacts.spec.ts:
```diff
@@ -165,14 +172,127 @@ describe('modules/manager/composer/artifacts', () => {
+    ]);
+  });
+
+  it('github hostRule for github.com with x-access-token set github-token in COMPOSER_AUTH', async () => {
```
new test

chrillep commented:
ping :)  . artifacts failing resolving it by manually running docker container with auth.json

rarkins commented:
Do we know for sure what caused it to fail for others last time, and therefore we can be confident that we're avoiding the problem this time?

etremblay commented:
> Do we know for sure what caused it to fail for others last time, and therefore we can be confident that we're avoiding the problem this time?

Am afraid it's difficult to know for shure.

The code modified 3 things

1) Understand `hostRules` with the `git-tags` `hostType` and add it in composer auth.json

Users in #17778 did not mention having such rules.  For this to be an issue they would need to hava a broken token in a `git-tags` host rule and it's not very likely.

2) We removed the `.replace('x-access-token:', '')`

This is what I fix here. What I understood form @viceice comment is that `x-access-token` is possibly added in Github App installations.

3) We accept only tokens that begins with the `ghp_` prefix

We added this condition because github now only accept personnal access token

> Support for password authentication was removed on August 13, 2021. Please use a personal access token instead.

Maybe these users have a github installation that still accept non-personnal access token.  We can remove this check to improve our chances of not doing a breaking change.  

Note that if it's the case, they will probably eventually have to update their tokens for PAT enventually anyway.

rarkins commented:
> 2. We removed the `.replace('x-access-token:', '')`
> 
> This is what I fix here. What I understood form @viceice comment is that `x-access-token` is possibly added in Github App installations.

This could be the problem

> 3. We accept only tokens that begins with the `ghp_` prefix
> 
> We added this condition because github now only accept personnal access token

This could also be the problem. The app uses `ghs_` tokens (designates an "app" token). Can we skip the check altogether?

etremblay commented:
> > 2. We removed the `.replace('x-access-token:', '')`
> > 
> > This is what I fix here. What I understood form @viceice comment is that `x-access-token` is possibly added in Github App installations.
> 
> This could be the problem
> 
> > 3. We accept only tokens that begins with the `ghp_` prefix
> > 
> > We added this condition because github now only accept personnal access token
> 
> This could also be the problem. The app uses `ghs_` tokens (designates an "app" token). Can we skip the check altogether?

Yes we can skip the check and the fix will still work for us.  If you set a bad token in your hostRule composer will just fail.

I removed the check but kept the documentation saying you should user a PAT.

etremblay commented:
> what about add a debug message when it's not. a pat or app token?

Good idea,  I improved the personal access token logic to prefer `ghp_` tokens but still use any token available if no `ghp_` is set.  In that case I log a debug.


https://github.com/renovatebot/renovate/pull/18004/commits/725ee9495d823e6ca21d007f9abc1a586758c38a

etremblay commented on lib/modules/manager/composer/artifacts.ts:
```diff
@@ -19,27 +19,44 @@ import {
+    gitTagsGithubToken
+  );
+  if (selectedGithubToken) {
+    if (!isGithubPersonalAccessToken(selectedGithubToken)) {
```
The whole point of the original pull request is that github does not accept `ghs_` token anymore.

```
In Git.php line 471:
                                                                               
  Failed to execute git clone --mirror -- 'https://ghs...IQn:x-oauth-basic@gi  
  thub.com/kronostechnologies/account-service-sdk-php.git' '/tmp/renovate-cac  
  he/others/composer/vcs/git-github.com-kronostechnologies-account-service-sd  
  k-php.git/'                                                                  
                                                                               
  Cloning into bare repository '/tmp/renovate-cache/others/composer/vcs/git-g  
  ithub.com-kronostechnologies-account-service-sdk-php.git'...                 
  remote: Support for password authentication was removed on August 13, 2021.  
  remote: Please see https://docs.github.com/en/get-started/getting-started-w  
  ith-git/about-remote-repositories#cloning-with-https-urls for information o  
  n currently recommended modes of authentication.                             
  fatal: Authentication failed for 'https://github.com/kronostechnologies/acc  
  ount-service-sdk-php.git/'                   
  ```

viceice replied:
if GitHub won't accept app token, the hosted app would never work.

so it must be some other issue. maybe its because the token is passed as username instead as password. that seems to be the issue of the log above.

BeyerJC replied:
> maybe its because the token is passed as username instead as password. 

Its the same for other applications. Username has to be set but is not used, i can be "Dummy" and Token is passed as password.

viceice replied:
sure, but the error above clearly has the opposite. the token is used as username.

etremblay replied:
https://github.com/composer/composer/issues/11064

etremblay replied:
That will be difficult to test without the risk of impacting users that had issues with the first fix https://github.com/renovatebot/renovate/issues/17778

chrillep replied:
> It seems application token could work with this composer setting
> 
> ```
> {
>   "http-basic": {
>     "github.com": { 
>       "username": "x-access-token", 
>       "password": "ghs_XXXXX"
>     }
>   }
> }
> ```

so how would one use ghs_XXXXX

remove git-tags setting ?

chrillep replied:
or how do i test the changes in this PR ? :)

etremblay replied:
This is the official project documentation https://github.com/renovatebot/renovate/blob/main/docs/development/local-development.md

I found it difficult to test the exact case of a github application.

chrillep replied:
@etremblay ty. but thats to much "Wall of Text" for me today :) 
ill keep waiting :)

chrillep reviewed (dismissed):
i approve :)

chrillep commented:
BTW does having in no-api in composer.json for your private repo have any effect on this issue?

https://getcomposer.org/doc/05-repositories.md


> If you set the no-api key to true on a github repository it will clone the repository as it would with any other git repository instead of using the GitHub API. But unlike using the git driver directly, Composer will still attempt to use github's zip files.

chrillep commented:
> > BTW does having in no-api in composer.json for your private repo have any effect on this issue?
> > https://getcomposer.org/doc/05-repositories.md#git-alternatives
> > > If you set the no-api key to true on a github repository it will clone the repository as it would with any other git repository instead of using the GitHub API. But unlike using the git driver directly, Composer will still attempt to use github's zip files.
> 
> Before all this, most of our private repositories had the noapi + type=github and we had the "Support for password authentication was removed" error. Moreover renovate could not list dependencies versions.
> 
> Now all our repos use type=vcs + https url ans renovate is able to fetch dependency versions and generate pull request. The only problem is with the composer artifact update (done by composer itself not the renovate codebase). That's what this pull request fix.

@etremblay alright we use vcs and no-api.

and also thanks for creating the PR.

viceice commented:
we probably also need to handle the new PAT type soon

https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/

etremblay commented:
> we probably also need to handle the new PAT type soon
> 
> https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/

Indeed.  With this version of the code we would only log a debug saying it's not a PAT. 

I just generated a token and it's prefixed by `github_pat_`.  I could update this pr to handle it like a `ghp`

chrillep commented:
so @viceice what are our options in out repos?
keep running via docker locally

1. checkout branch.
2. make sure you have a correct auth.json in the root of the repo
3. commit lockfile
4. push
5. wait for all CI stuff 
6. merge 
7. go to next dep PR
8. update said PR
9. repeat

rarkins commented on lib/modules/manager/composer/artifacts.ts:
```diff
@@ -19,27 +19,44 @@ import {
+    gitTagsGithubToken
+  );
+  if (selectedGithubToken) {
+    if (!isGithubPersonalAccessToken(selectedGithubToken)) {
+      logger.debug(`Selected github token is not a Personal Access Token`);
+    }
```
```suggestion
    if (isGithubPersonalAccessToken(selectedGithubToken)) {
      logger.debug('Using GitHub Personal Access Token');
    } else if (isGithubServerToServerToken(selectedGithubToken)) {
      logger.debug('Using GitHub Server-to-Server token');
    } else {
      logger.debug('Using unknown GitHub token type');
    }
```

etremblay replied:
Good idea, I will add the fine grained token too

rarkins commented on lib/modules/manager/composer/utils.ts:
```diff
@@ -109,3 +111,30 @@ export function extractConstraints(
+export function isGithubPersonalAccessToken(token: string): boolean {
+  return regEx(/^ghp_/).test(token);
+}
+
```
```suggestion

export function isGithubServerToServerToken(token: string): boolean {
  return regEx(/^ghs_/).test(token);
}
```

etremblay commented:
Added @rarkins suggestion + support for Fine-grainted token + update with main branch

HonkingGoose commented on docs/usage/getting-started/private-packages.md:
```diff
@@ -179,6 +179,21 @@ The following details the most common/popular manager artifacts updating and how
 Any `hostRules` token for `github.com` or `gitlab.com` are found and written out to `COMPOSER_AUTH` in env for Composer to parse.
 Any `hostRules` with `hostType=packagist` are also included.
+For dependencies on `github.com` without a packagist server, a hostRule with `hostType=git-tags` should be used with a personal access token (not an application token).
+Do not add a hostRule with `hostType=github` because it can override the default renovate application token for everything else and cause unwanted side effects.
```
```suggestion
Avoid adding a `hostRule` with `hostType=github` because:

- it overrides the default Renovate application token for everything else 
- it causes unwanted side effects.
```

HonkingGoose commented on docs/usage/getting-started/private-packages.md:
```diff
@@ -179,6 +179,21 @@ The following details the most common/popular manager artifacts updating and how
+For dependencies on `github.com` without a packagist server, a hostRule with `hostType=git-tags` should be used with a personal access token (not an application token).
+Do not add a hostRule with `hostType=github` because it can override the default renovate application token for everything else and cause unwanted side effects.
+
+The repository in `composer.json` should have the `vcs` type with a `https` url. Ex:
```
```suggestion
The repository in `composer.json` should have the `vcs` type with a `https` URL.
For example:
```

HonkingGoose commented on docs/usage/getting-started/private-packages.md:
```diff
@@ -179,6 +179,21 @@ The following details the most common/popular manager artifacts updating and how
 
 Any `hostRules` token for `github.com` or `gitlab.com` are found and written out to `COMPOSER_AUTH` in env for Composer to parse.
 Any `hostRules` with `hostType=packagist` are also included.
+For dependencies on `github.com` without a packagist server, a hostRule with `hostType=git-tags` should be used with a personal access token (not an application token).
```
```suggestion
For dependencies on `github.com` without a Packagist server: use a Personal Access Token for `hostRule` with `hostType=git-tags`, do not use an application token.
```

HonkingGoose commented on docs/usage/getting-started/private-packages.md:
```diff
@@ -179,6 +179,25 @@ The following details the most common/popular manager artifacts updating and how
+Avoid adding a `hostRule` with `hostType=github` because:
+
+- it overrides the default Renovate application token for everything else 
+- it causes unwanted side effects.
```
```suggestion
- it causes unwanted side effects
```

Whoops, we don't use punctuation at the end of a list item. 🙂

chrillep commented:
Tralla öppning forward to this ! 😀

etremblay commented:
Updated with master and fixed code coverage

rarkins commented:
Needs conflict resolution

etremblay commented:
> Needs conflict resolution

Fixed

viceice commented on lib/modules/manager/composer/utils.ts:
```diff
@@ -109,3 +111,49 @@ export function extractConstraints(
+}
+
+export function isGithubPersonalAccessToken(token: string): boolean {
+  return regEx(/^ghp_/).test(token);
```
better to use tring.startsWith, should be faster than regex.

viceice commented on lib/modules/manager/composer/artifacts.ts:
```diff
@@ -18,26 +18,51 @@ import {
+    gitTagsGithubToken
+  );
+  if (selectedGithubToken) {
+    if (isGithubPersonalAccessToken(selectedGithubToken)) {
+      logger.debug('Using GitHub Personal Access Token');
+    } else if (isGithubServerToServerToken(selectedGithubToken)) {
+      logger.debug('Using GitHub Server-to-Server token');
+    } else if (isGithubFineGrainedPersonalAccessToken(selectedGithubToken)) {
+      logger.debug('Using GitHub Fine-grained Personal Access Token');
+    } else {
+      logger.debug('Using unknown GitHub token type');
+    }
```
i think those log messages should be moved to the  take personal access token function.

etremblay replied:
Done, it allowed a more precise logging and simplified the unit tests to get code coverage on those lines.

Thanks.

viceice reviewed (changes requested):
see regex issue

rarkins commented:
This branch should automerge the next time it (a) is up to date with main branch, and (b) has passed all tests
